lib/Core/Search/Common/Messenger/MessageHandler/Search/Content/CopyContentHandler.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Netgen\IbexaSearchExtra\Core\Search\Common\Messenger\MessageHandler\Search\Content;
+
+use Ibexa\Contracts\Core\Persistence\Content\Handler as ContentHandler;
+use Ibexa\Contracts\Core\Persistence\Content\Location\Handler as LocationHandler;
+use Ibexa\Contracts\Core\Repository\Exceptions\NotFoundException;
+use Ibexa\Contracts\Core\Search\VersatileHandler as SearchHandler;
+use Netgen\IbexaSearchExtra\Core\Search\Common\Messenger\Message\Search\Content\CopyContent;
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+
+use function sprintf;
+
+final class CopyContentHandler
+{
+    public function __construct(
+        private readonly ContentHandler $contentHandler,
+        private readonly LocationHandler $locationHandler,
+        private readonly SearchHandler $searchHandler,
+        private readonly LoggerInterface $logger = new NullLogger(),
+    ) {}
+
+    public function __invoke(CopyContent $message): void
+    {
+        try {
+            $content = $this->contentHandler->load(
+                $message->contentId,
+                $message->versionNo,
+            );
+        } catch (NotFoundException) {
+            $this->logger->info(
+                sprintf(
+                    '%s: Content #%d in version %d is gone, aborting',
+                    $this::class,
+                    $message->contentId,
+                    $message->versionNo,
+                ),
+            );
+
+            return;
```

Comment:
Maybe instead of returning, you can use `UnrecoverableMessageHandlingException`. That way we ensure the message is not retried, and it ends up in failure transport for easier debugging.

Reply:
The message will not be retried because of that `return`, it's part of the normal operation. As far as the mechanism knows, it was handled successfully. It's a actually a fully legitimate case, as there might be a big delay between recording and handling a message, so we can't count on the data being valid. For example the Content might have been deleted in the meantime, and there is nothing to do here because the other message will do it's job.

Reply:
Delete is normal operation, I agree, others are not, and it would be nice to have the entire stack trace in failure transport if needed, that:s why I prefer the exception.

Reply:
Discussed this in person, all cases where exceptions are caught handle the possible deletion between creation and handling of a message and are part of normal operation.